system_tests/stylus_trace_test.go

```diff
@@ -0,0 +1,483 @@
+	create2Args = append(create2Args, deployCode...)
+	result = sendAndTraceTransaction(t, builder, program, startValue, create2Args)
+	checkOpcode(t, result, 10, vm.CREATE2, startValue.Bytes(), nil, intToBytes(len(deployCode)), salt[:])
+	checkOpcode(t, result, 11, vm.POP, create2Addr[:])
+}
+
+// TestStylusTraceEquivalence compares a Stylus trace with a equivalent Solidity/EVM trace. Notice
+// the Stylus trace does not contain all opcodes from the Solidity/EVM trace. Instead, this test
+// only checks that both traces contain the same basic opcodes.
+func TestStylusTraceEquivalence(t *testing.T) {
+	const jit = false
+	builder, auth, cleanup := setupProgramTest(t, jit)
+	ctx := builder.ctx
+	l2client := builder.L2.Client
+	defer cleanup()
+
+	// Args for storing and loading from storage
+	const (
+		storageKind = 0x10
+		storeAction = storageKind | 0x00
+		loadAction  = storageKind | 0x01
+		logModifier = 0x08
+	)
+	key := testhelpers.RandomHash()
+	value := testhelpers.RandomHash()
+	args := []byte{2} // number of actions
+	// first action
+	args = binary.BigEndian.AppendUint32(args, 1+64) // length
+	args = append(args, storeAction|logModifier)
+	args = append(args, key.Bytes()...)
+	args = append(args, value.Bytes()...)
+	// second action
+	args = binary.BigEndian.AppendUint32(args, 1+32) // length
+	args = append(args, loadAction|logModifier)
+	args = append(args, key.Bytes()...)
+
+	// Trace recursive call in wasm
+	wasmMulticall := deployWasm(t, ctx, auth, l2client, rustFile("multicall"))
+	colors.PrintGrey("wasm multicall deployed at ", wasmMulticall)
+	wasmArgs := argsForMulticall(vm.CALL, wasmMulticall, nil, args)
+	wasmResult := sendAndTraceTransaction(t, builder, wasmMulticall, nil, wasmArgs)
+
+	// Trace recursive call in evm
+	evmMulticall, tx, _, err := mocksgen.DeployMultiCallTest(&auth, builder.L2.Client)
+	Require(t, err)
+	_, err = EnsureTxSucceeded(ctx, l2client, tx)
+	Require(t, err)
+	colors.PrintGrey("evm multicall deployed at ", evmMulticall)
+	evmArgs := argsForMulticall(vm.CALL, evmMulticall, nil, args)
+	evmResult := sendAndTraceTransaction(t, builder, evmMulticall, nil, evmArgs)
+
+	// For some opcodes in the wasmTrace, make sure there is an equivalent one in the evmTrace.
+	argsLen := intToBytes(len(args))
+	offset := skipCheck
+	checkOpcode(t, wasmResult, 3, vm.CALL, skipCheck, wasmMulticall[:], nil, offset, argsLen, offset, nil)
+	checkOpcode(t, evmResult, 3120, vm.CALL, skipCheck, evmMulticall[:], nil, offset, argsLen, offset, nil)
+
+	checkOpcode(t, wasmResult, 5, vm.SSTORE, key[:], value[:])
+	checkOpcode(t, evmResult, 3905, vm.SSTORE, key[:], value[:])
+
+	topic := common.Hex2Bytes("6ab08a9a891703dcd5859f8e8328215fef6d9f250e7d58267bee45aabaee2fa8")
+	logLen := intToBytes(0x60)
+	checkOpcode(t, wasmResult, 6, vm.LOG1, offset, logLen, topic)
+	checkOpcode(t, evmResult, 3944, vm.LOG1, offset, logLen, topic)
+
+	checkOpcode(t, wasmResult, 7, vm.SLOAD, key[:])
+	checkOpcode(t, evmResult, 4645, vm.SLOAD, key[:])
```

Comment:
also add: RETURN
I would be happy to also see STOP but not sure the current EVM contract can do that (it can issue a return with no data but I think that would still be return)